aloha-core/src/main/scala/com/eharmony/aloha/semantics/MorphableSemantics.scala

```diff
@@ -0,0 +1,98 @@
+  *
+  * val xF = X[Float]
+  * val riD = implicitly[RefInfo[Double]]
+  * val xDo: Option[X[Double]] = xF.morph[Double]
+  * val xDo1: Option[X[Double]] = xF.morph(riD)
+  * }}}
+  *
+  * or use it generically:
+  *
+  * {{{
+  * def morph[M[_] <: Semantics[_], A, B: RefInfo](m: MorphableSemantics[M, A]): Option[Semantics[B]] = {
+  *   m.morph[B].map(ms => ms.semantics)
+  * }
+  *
+  * val sDO: Option[Semantics[Double]] = morph(xF)
+  * }}}
+  *
+  * @tparam M a type constructor providing the structure of the [[Semantics]].  This will most
+  *           likely always be the concrete subclass of [[Semantics]] that implements
+  *           [[MorphableSemantics]].
+  * @tparam A the type parameter to which `M` is applied for this instance of [[Semantics]].
+  *           By the self-type, also the type parameter for the [[Semantics]].
+  * @author deaktator
+  */
+trait MorphableSemantics[M[_] <: Semantics[_], A] { self: Semantics[A] =>
+
+  /**
+    * Get `this` instance as a `Semantics[A]`.
+    *
+    * This could be implemented fully here because the self-type allows it, but it is left for
+    * the derived classes because they can return a more specific type.  For instance:
+    *
+    * {{{
+    * case class X[A]() extends Semantics[A] with MorphableSemantics[X, A] {
+    *   def semantics: X[A] = this
+    *   def morph[B: RefInfo]: Option[X[B]] = Some(X[B]())
+    * }
+    * }}}
+    *
+    * can be used generically, in which case a value of type `Semantics[A]` is returned in cases
+    * like the following:
+    *
+    * {{{
+    * def generic[M[_] <: Sem[_], A](m: Morphable[M, A]): Sem[A] = m.semantics
+    * }}}
+    *
+    * In cases where the concrete type is known, a more specific type can be return, e.g.:
+    *
+    * {{{
+    * def specific[A](x: X[A]): X[A] = x.semantics
+    * }}}
+    * @return `this`, cast as `Semantics[A]`.
+    */
+  def semantics: Semantics[A]
+
+  /**
+    * Attempt to morph this [[MorphableSemantics]] into a new one with the same structure.
+    *
+    * The new [[MorphableSemantics]] will have the same structure (as dictated by the `M`
+    * type parameter) but with a different second parameter.  This new instance must also be
+    * a `Semantics[B]` because of the type bound on `M` and the self-type of
+    * [[MorphableSemantics]].  It can be the case that a specific implementation of
+    * [[MorphableSemantics]] doesn't allow for all `B`, in which case a `None` can be returned.
+    * @param ri reflection information that may be necessary to determine whether to create
+    *           the [[MorphableSemantics]] that was requested.
+    * @tparam B input type for the new [[MorphableSemantics]] instance that might be created.
+    *           A [[MorphableSemantics]] instance may choose not allow morphing to all `B`.
+    *           In that case, a `None` will be returned.
+    * @return
+    */
+  def morph[B](implicit ri: RefInfo[B]): Option[MorphableSemantics[M, B]]
```

Comment:
Should this use context bounds instead of an implicity `RefInfo[B]` parameter?  Who has any thoughts?